### PR TITLE
Add RMSEs and targets

### PR DESCRIPTION
**Pull request by SvenKlaassen**
### Description
Add RMSE evaluations for each nuisance component to the models. The RMSEs can be accessed through the method `.rmses` and are added to the `summary()`.
Further, the targets for each component can be accessed through the method `.nuisance_targets`, which returns a dictionary which contains the nuisance targets for each nuisance component (as an array for each repetition and each coefficient). 
The new method `evaluate_learners` for `DoubleML` objects allows to evaluate the nuisance learners for a callable metric which is based on vector-based (shape of `(1, n)`) inputs `y_pred` and `y_true` (see e.g. [scikit-learn](https://scikit-learn.org/stable/modules/model_evaluation.html#regression-metrics)).

## Notes
Fix a bug to correctly save models and predictions for the PLIV model #184 

### PR Checklist
Please fill out this PR checklist (see our [contributing guidelines](https://github.com/DoubleML/doubleml-for-py/blob/main/CONTRIBUTING.md#checklist-for-pull-requests-pr) for details).

- [x] The title of the pull request summarizes the changes made.
- [x] The PR contains a detailed description of all changes and additions.
- [x] References to related issues or PRs are added.
- [x] The code passes all (unit) tests.
- [x] Enhancements or new feature are equipped with unit tests.
- [x] The changes adhere to the PEP8 standards.

### CODE REVIEW

File: doubleml/double_ml.py [hunk 1]
```diff
@@ -483,8 +507,10 @@ def fit(self, n_jobs_cv=None, store_predictions=False, store_models=False):
 
                 self._set_score_elements(score_elements, self._i_rep, self._i_treat)
 
+                # calculate rmses and store predictions and targets of the nuisance models
+                self._calc_rmses(preds['predictions'], preds['targets'])
```
Comment (PhilippBach) [inline doubleml/double_ml.py:511]: Hi @SvenKlaassen , 

thanks for this PR! I think adding some diagnostics is really nice.

I think it should be possible to make this a bit more general by using (maybe only a subset of) sklearn's [metrics](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics) either by letting users pass a `callable` for evaluation of the nuisance predictions or by directly supporting the measures. For example, in case $D$ or $Y$ are binary, classification error or cross-entropy loss might be more relevant than the RMSE. What do you think about this?

Comment (PhilippBach) [reply doubleml/double_ml.py:511]: I guess that sklearn's measures have built in some methods to handle exceptions that might occur ...

File: doubleml/double_ml.py [hunk 2]
```diff
@@ -434,7 +456,7 @@ def __psi_deriv(self):
     def __all_se(self):
         return self._all_se[self._i_treat, self._i_rep]
 
-    def fit(self, n_jobs_cv=None, store_predictions=False, store_models=False):
+    def fit(self, n_jobs_cv=None, store_predictions=True, store_models=False):
```
Comment (PhilippBach) [inline doubleml/double_ml.py:459]: Thanks @SvenKlaassen  - I think it's a good idea to set the default for `store_predictions` to `True`  👍

File: doubleml/double_ml.py [hunk 3]
```diff
@@ -990,22 +1017,103 @@ def _initialize_boot_arrays(self, n_rep_boot):
...
+        >>> from doubleml.datasets import make_irm_data
+        >>> from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
+        >>> np.random.seed(3141)
+        >>> ml_g = RandomForestRegressor(n_estimators=100, max_features=20, max_depth=5, min_samples_leaf=2)
+        >>> ml_m = RandomForestClassifier(n_estimators=100, max_features=20, max_depth=5, min_samples_leaf=2)
+        >>> data = make_irm_data(theta=0.5, n_obs=500, dim_x=20, return_type='DataFrame')
+        >>> obj_dml_data = dml.DoubleMLData(data, 'y', 'd')
+        >>> dml_irm_obj = dml.DoubleMLIRM(obj_dml_data, ml_g, ml_m)
+        >>> dml_irm_obj.fit()
+        >>> dml_irm_obj.evaluate_learners(metric=mean_absolute_error)
```
Comment (PhilippBach) [inline doubleml/double_ml.py:1084]: I'm afraid we get a little problem here if we use callables for classification here, like for example if we run instead:

```
import numpy as np
import doubleml as dml
from sklearn.metrics import log_loss
from doubleml.datasets import make_irm_data
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
np.random.seed(3141)
ml_g = RandomForestRegressor(n_estimators=100, max_features=20, max_depth=5, min_samples_leaf=2)
ml_m = RandomForestClassifier(n_estimators=100, max_features=20, max_depth=5, min_samples_leaf=2)
data = make_irm_data(theta=0.5, n_obs=500, dim_x=20, return_type='DataFrame')
obj_dml_data = dml.DoubleMLData(data, 'y', 'd')
dml_irm_obj = dml.DoubleMLIRM(obj_dml_data, ml_g, ml_m)
dml_irm_obj.fit()
dml_irm_obj.evaluate_learners(metric=log_loss)
```

File: doubleml/_utils.py [hunk 4]
```diff
@@ -150,7 +151,9 @@ def _dml_cv_predict(estimator, x, y, smpls=None,
                                      for idx, (train_index, test_index) in enumerate(smpls))
 
         preds = np.full(n_obs, np.nan)
+        targets = np.full(n_obs, np.nan)
         train_preds = list()
+        train_targets = list()
```
Comment (PhilippBach) [inline doubleml/_utils.py:156]: As far as I see, we're not using the `train_targets` right now, right? I guess that's alright and we can leave a in-sample (fold-wise cross-validated) vs. out-of-sample (cross-fitted) evaluation for later. I'm wondering how we can implement this in a clever way... 🤔

File: doubleml/double_ml.py [hunk 5]
```diff
@@ -990,22 +1017,103 @@ def _initialize_boot_arrays(self, n_rep_boot):
...
+         'ml_g1': array([[0.91659939]]),
+         'ml_m': array([[0.36350912]])}
+        """
+        # if no learners are provided try to evaluate all learners
+        if learners is None:
+            learners = self.params_names
+
+        # check metric
+        if not callable(metric):
+            raise TypeError('metric should be either a callable. '
```
Comment (PhilippBach) [inline doubleml/double_ml.py]: I fixed the message in https://github.com/DoubleML/doubleml-for-py/commit/10b180e9f2bf4cc9c199f1064ba4b92ef481988d

**Comment by PhilippBach** [comment]
We should also demonstrate the use of the new feature(s) in a short example, see https://github.com/DoubleML/doubleml-docs/issues/114
